igvc_navigation/src/mapper/octomapper.cpp

```diff
@@ -0,0 +1,215 @@
+{
+  return 1 - (1 / (1 + exp(log_odds)));
+}
+
+std::pair<int, int> Octomapper::toMapCoordinates(double x, double y) const
+{
+  int x_map = static_cast<int>((map_options_.length / 2.0 + x) / octree_options_.resolution);
+  int y_map = static_cast<int>((map_options_.width / 2.0 + y) / octree_options_.resolution);
+  return std::make_pair(x_map, y_map);
+}
+
+void Octomapper::get_updated_map(struct pc_map_pair &pc_map_pair) const
+{
+  if (pc_map_pair.map == nullptr)
+  {
+    create_map(pc_map_pair);
+  }
+
+  // Traverse entire tree
+  std::vector<std::vector<float>> odds_sum(
+      static_cast<unsigned long>(map_options_.lengthGrid()),
+      std::vector<float>(static_cast<unsigned long>(map_options_.widthGrid()), map_options_.default_logodds));
+  for (octomap::OcTree::iterator it = pc_map_pair.octree->begin(), end = pc_map_pair.octree->end(); it != end; ++it)
+  {
+    // If this is a leaf at max depth, then only update that node
+    if (it.getDepth() == pc_map_pair.octree->getTreeDepth())
+    {
+      auto [x, y] = toMapCoordinates(it.getX(), it.getY());
+      if (x < map_options_.lengthGrid() && y < map_options_.widthGrid())
+      {
+        odds_sum[x][y] += it->getLogOdds();
+      }
+      else
+      {
+        ROS_ERROR_STREAM_THROTTLE_NAMED(10, "Point Outside", "Point outside!");
+      }
+    }
+    else
+    {
+      // This isn't a leaf at max depth. Time to iterate
+      int grid_num = 1u << (pc_map_pair.octree->getTreeDepth() - it.getDepth());
+      auto [x, y] = toMapCoordinates(it.getX(), it.getY());
+      for (int dx = 0; dx < grid_num; dx++)
+      {
+        for (int dy = 0; dy < grid_num; dy++)
+        {
+          if (x < map_options_.lengthGrid() && y < map_options_.widthGrid())
+          {
+            odds_sum[x + dx][y + dy] += it->getLogOdds();
+          }
+          else
+          {
+            ROS_ERROR_STREAM_THROTTLE_NAMED(10, "Point Outside", "Point outside!");
+          }
+        }
+      }
+    }
+  }
+
+  // Transfer from log odds to normal probability
+  for (int i = 0; i < map_options_.lengthGrid(); i++)
+  {
+    for (int j = 0; j < map_options_.widthGrid(); j++)
+    {
+      pc_map_pair.map->at<uchar>(i, j) = toCharProb(fromLogOdds(odds_sum[i][j]));
+    }
+  }
+}
+
+void Octomapper::create_map(pc_map_pair &pair) const
+{
+  int length = static_cast<int>(map_options_.lengthGrid());
+  int width = static_cast<int>(map_options_.widthGrid());
+  pair.map = boost::make_shared<cv::Mat>(length, width, map_encoding_, 127);
+}
+
+void Octomapper::insertScan(const tf::Point &sensor_pos, struct pc_map_pair &pair, const PointCloud &pc,
+                            ProbabilityModel model, double range) const
+{
+  double old_prob_hit = pair.octree->getProbHit();
+  double old_prob_miss = pair.octree->getProbMiss();
+
+  pair.octree->setProbHit(model.prob_hit);
+  pair.octree->setProbMiss(model.prob_miss);
+
+  octomap::point3d sensor = octomap::pointTfToOctomap(sensor_pos);
+  sensor.z() = 0;
```

Comment:
what's the justification for setting the sensor height to 0 when inserting a ray into the map?

Reply:
see above.